cmake/Platform/Core/Targets/ArduinoFirmwareTargetCreator.cmake

```diff
@@ -20,18 +20,37 @@ function(create_arduino_firmware_target TARGET_NAME BOARD_ID ALL_SRCS ALL_LIBS
 
     string(STRIP "${ALL_SRCS}" ALL_SRCS)
     add_executable(${TARGET_NAME} "${ALL_SRCS}")
+    if(ARDUINO_CMAKE_GENERATE_SHARED_LIBRARIES)
+        add_library(${TARGET_NAME} SHARED "${ALL_SRCS}")
+    else()
+        add_executable(${TARGET_NAME} "${ALL_SRCS}")
+    endif()
     set_target_properties(${TARGET_NAME} PROPERTIES SUFFIX ".elf")
 
     set_board_flags(ARDUINO_COMPILE_FLAGS ARDUINO_LINK_FLAGS ${BOARD_ID} ${MANUAL})
 
     set_target_properties(${TARGET_NAME} PROPERTIES
             COMPILE_FLAGS "${ARDUINO_COMPILE_FLAGS} ${COMPILE_FLAGS}"
             LINK_FLAGS "${ARDUINO_LINK_FLAGS} ${LINK_FLAGS}")
-    target_link_libraries(${TARGET_NAME} ${ALL_LIBS} "-lc -lm")
 
+    if(ARDUINO_CMAKE_GENERATE_SHARED_LIBRARIES)
+      # When building a shared library we must make sure that
+      # all symbols from the intermediate static libraries end up in the
+      # static library
+      #
+      target_link_libraries(${TARGET_NAME} PUBLIC "-Wl,--whole-archive" ${ALL_LIBS} "-Wl,--no-whole-archive")
```

Comment:
There are undefined symbols, if we don not link the whole intermediate archives. For virtual builds, we do not want any dead-code elimination.